### PR TITLE
WIP: cre technical patterns

### PR DESCRIPTION
**Pull request by Ottovsky**
Please treat it as an invitation to the discussion in comments, I think there is still some work to be done before merging. 

Let's agree whether those patterns fit the CRE idea and we could possibly support them in the future. Also if you have any comments regarding content, formatting, typos, please don't hesitate to raise them.

### CODE REVIEW

**Comment by darren-reddick** [comment]
Could we put a guide on running this locally? I've been running jekyll locally and its nice to have the option to see the changes rendered.

**Comment by cewood** [comment]
> Could we put a guide on running this locally? I've been running jekyll locally and its nice to have the option to see the changes rendered.

Maybe we could target migrating this to Hugo later on as well, if anyone is interested? There's no reason to be migrating sites to Jekyll in 2020, just more cargo culting :man_facepalming:

**Comment by darren-reddick** [comment]
> Maybe we could target migrating this to Hugo later on as well, 

Yes. That could be a good thing to do. 
How is this thing actually hosted atm anyway?

**Comment by cewood** [comment]
> > Maybe we could target migrating this to Hugo later on as well,
> 
> Yes. That could be a good thing to do.

Kewl, well we can add an item to our backlog if there's interest to track doing that later.


> How is this thing actually hosted atm anyway?

It's being published to GitHub pages, the hint is here https://github.com/ContainerSolutions/cloud-native-patterns/blob/master/CNAME

That could also be something we considering changing, moving it to Netlify or similar, so we can get viewable deployment previews on PR's, to properly see things before merging or releasing, etc. I've had great success using Cloudflare, Netlify, and Hugo before, super fast free hosting and nice experience overall.

**Comment by Ottovsky** [comment]
> > Could we put a guide on running this locally? I've been running jekyll locally and its nice to have the option to see the changes rendered.
> 
> Maybe we could target migrating this to Hugo later on as well, if anyone is interested? There's no reason to be migrating sites to Jekyll in 2020, just more cargo culting man_facepalming

I like the idea, I will create a low priority task in the jira for it.

File: _technical/cicd-as-platform.md [hunk 1]
```diff
@@ -4,52 +4,43 @@ description: The CI/CD system is treated as a platform on top of the cloud platf
...
 CI/CD services on an Enterprise scale are difficult to implement. Taking into account 
 requirements for auditing, compliance and team flexibility will lead to a system
 which is customized to a high degree. Often the responsibility for the these services
 end up with the Platform Team, which is managing infrastructure. While CI/CD is related
 to the infrastructure platform, it has it's own goals and a different domain (in DDD terms).
 
-## Solution
+## Therefore
 
 Create a dedicated product team, which will build and manage CI/CD services. This team will
```
Comment (darren-reddick) [inline _technical/cicd-as-platform.md:28]: Having the word "product" here is confusing. Would it not read better if it just said "create a dedicated team" as a reference to the CI/CD Platfrom team.

File: _technical/gitops.md [hunk 2]
```diff
@@ -1,44 +1,40 @@
...
 title: GitOps
-description: Description part 1
+description: GitOps - a single source of truth for infrastructure configuration
 layout: pattern
 ---
 
 ![GitOps]({{ site.baseurl }}/assets/images/Generic.png)
 
-Description part 2
+You have setup all your infrastructure as code (see #InfrastructureAsCode pattern).
```
Comment (darren-reddick) [inline _technical/gitops.md]: I dont see an InfrastructureAsCode pattern anywhere in this repo. I am going to remove this reference for now....

File: technical.md [hunk 3]
```diff
@@ -1,6 +1,6 @@
 ---
 title: Technical Patterns
-description: "Yadayada."
+description: "Patterns that focus on how to manage and operate Cloud Native services whilst ensuring their reliability and security."
```
Comment (darren-reddick) [inline technical.md]: I have tried to add a description but its a work in progress (and def better than "yadayada").

Would def appreciate input from others to try and get it to a sentence that encompasses what these patterns are about.

**Comment by darren-reddick** [comment]
Im not sold on the "structured-logging" pattern - i'm not sure what it is advocating: is it the centralization of the logs? Is it a particular format for the logs? Of course these are good things but I dont think the case for them is well presented here. 
The other patterns present a "scenario > problem > solution" structure which I think is missing here so I think we have more work to do on this one if we are going to keep it.

**Comment by Ottovsky** [comment]
Agree, I was not bought by this when I refactored it. I think it is about having logs structured and serialized in a way that it is easy to store and query in ELK or something similar. The reference to the RFC is not very clear to me, I would just drop it for now.

File: _technical/cicd-as-platform.md (hunk 1 above)
Comment (Ottovsky) [reply _technical/cicd-as-platform.md:28]: You are right, without defining what does the "product team" means, we should not use it here.

File: _technical/structured-logging.md [hunk 4]
```diff
@@ -1,44 +1,35 @@
 ---
```
Comment (Ottovsky) [inline _technical/structured-logging.md]: As per @darren-reddick comment, what does the rest of the @ContainerSolutions/cre thinks about this technical pattern? Should we try to refactor it or just get rid of it for the time being?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+---
+title: <name of the pattern>
+description: <description>
+layout: pattern
+---
+
+![<name of the pattern>]({{ site.baseurl }}/assets/images/generic.png) //link to image if relevant
+
+<situation description>
```
Comment (darren-reddick) [inline CONTRIBUTING.md]: I think this should actually be <context> as mentioned in the Confluence doc

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+## In This Context
+
+<set context>
```
Comment (darren-reddick) [inline CONTRIBUTING.md]: This section actually describes the problems and pitfalls which are found within this context. So maybe <problems within context>?

**Comment by darren-reddick** [comment]
> I totally forgot about that :( I will adjust it accordingly

Here you go.... https://container-solutions.atlassian.net/wiki/spaces/CRE/pages/263356540/Tech+Patterns+Cloud+Native+Patterns+Discussion

File: technical.md (hunk 3 above)
Comment (cewood) [reply technical.md]: Looks pretty close to me. Is this an improvement?

```suggestion
description: "Patterns that focus on the technical and architectural aspects of how to build and operate Cloud Native services whilst ensuring their reliability and security."
```

File: _technical/structured-logging.md (hunk 4 above)
Comment (cewood) [reply _technical/structured-logging.md]: Structured logging is definitely legitimate thing, and refers to the tokenisation of logs. So in terms of standard fields, and also in terms of formatting, whether that be JSON or comma separated with a fixed field order, etc...

File: technical.md (hunk 3 above)
Comment (darren-reddick) [reply technical.md]: Thanks. That is better now.

File: _technical/structured-logging.md (hunk 4 above)
Comment (darren-reddick) [reply _technical/structured-logging.md]: Yes. I dont think the argument is against the pattern but that it isnt presented in a particularly clear way here (for me anyway). I think this needs a fair bit of work to get into a better state.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+- Becoming a maintainer
+
+## We Develop with Github
+We use github to host code, to track issues and feature requests, as well as accept pull requests.
+
+## We Use [Github Flow](https://guides.github.com/introduction/flow/index.html), So All Code Changes Happen Through Pull Requests
+Pull requests are the best way to propose changes to the patterns. We actively welcome your pull requests:
+
+1. Fork the repo and create your branch from `master`.
+2. If you've added pattern, double check for typos
```
Comment (rodrmartinez) [inline CONTRIBUTING.md]: 2. If you've added a  pattern, double check for typos

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+3. If you've changed pattern, explain change in the PR comment
```
Comment (rodrmartinez) [inline CONTRIBUTING.md]: If you've changed a pattern, explain your changes with a comment in the PR

File: _technical/cicd-as-platform.md [hunk 9]
```diff
@@ -4,52 +4,43 @@ description: The CI/CD system is treated as a platform on top of the cloud platf
...
 be a consumer of the infrastructure platform and a service provider towards development
 teams. The CI/CD Platform Team should treat CI/CD services as a product offered to all
-development teams. This means feature requests should be phrased in way that is useful to
+development teams. 
+
+This means feature requests should be phrased in way that is useful to
```
Comment (rodrmartinez) [inline _technical/cicd-as-platform.md]: This means feature requests should be phrased in a way that is useful to

File: _technical/dynamic-secrets.md [hunk 10]
```diff
@@ -1,28 +1,26 @@
...
 Building a distributed, microservices architecture. Many services require passwords, keys, tokens or other sensitive data which cannot be stored in plain text.
 
-## Problem
-
 Teams tend to end up with one of two sets of problems:
 
-### High Security - Low Agility
+##### High Security - Low Agility
 
 - All credentials are managed (manually) be a security team. Nobody usually knows where this team actually sits, there only interface is an obscure ticket filing system.
```
Comment (rodrmartinez) [inline _technical/dynamic-secrets.md]: - All credentials are managed (manually) be a security team. Nobody usually knows where this team actually sits, the only interface is an obscure ticket filing system.

File: _technical/ephemeral-environments.md [hunk 11]
```diff
@@ -1,25 +1,21 @@
...
-## Context
+![Ephemeral Environments]({{ site.baseurl }}/assets/images/cs.png)
 
 A team running applications on modern cloud native platform (such as Kubernetes) often requires a variety of environments. Getting a production like environment running on a developers laptop is often not possible. The benefit of automated infrastructure means the cost of creating these environments is cheap. Running and managing them not so much.
 
-## Problem
+## In This Context
 
-Cotnainer technology has taken us a long way to resolving the 'it works on my machine' problem. However, in a microservice environment, the number of dependencies for truly validating the functionality of service can still make this a big challenge.
+Container technology has taken us a long way to resolving the 'it works on my machine' problem. However, in a microservice environment, the number of dependencies for truly validating the functionality of service can still make this a big challenge.
```
Comment (rodrmartinez) [inline _technical/ephemeral-environments.md:13]: Container technology has taken us a long way to resolve the 'it works on my machine' problem. However, in a microservice environment, the number of dependencies for truly validating the functionality of a service can still make this a big challenge.

**Review by rodrmartinez — COMMENTED** [review]
Just some typos

File: _technical/ephemeral-environments.md (hunk 11 above)
Comment (darren-reddick) [reply _technical/ephemeral-environments.md:13]: I think you are commenting on the code pre-change here.

File: _technical/structured-logging.md (hunk 4 above)
Comment (darren-reddick) [reply _technical/structured-logging.md]: I have opened an issue to track this [here](https://github.com/ContainerSolutions/cloud-native-patterns/issues/10) and will remove the pattern from this PR.

**Review by cewood — APPROVED** [review]
LGTM :+1: 

@darren-reddick can you please review and merge if there's no further feedback.

File: _technical/cicd-as-platform.md [hunk 12]
```diff
@@ -4,52 +4,43 @@ description: The CI/CD system is treated as a platform on top of the cloud platf
...
-## Context
-
-Enterprise company is developing Cloud Native CI/CD services during a transormation or
+An Enterprise company is developing Cloud Native CI/CD services during a transformation or
 cloud migration. CI/CD services include building of container images, code quality checks,
 security scans and deployment of artifacts to environments. CI/CD plays a significant
 part in compliance requirements and making sure production systems are secure.
 
-## Problem
+![CI/CD as Platform]({{ site.baseurl }}/assets/images/cicd.png)
```
Comment (darren-reddick) [inline _technical/cicd-as-platform.md]: The image should be the first text after the meta-data section at the top to follow the standard pattern. This applies to all of the patterns.

**Review by darren-reddick — CHANGES_REQUESTED** [review]

File: _technical/cicd-as-platform.md (hunk 12 above)
Comment (darren-reddick) [reply _technical/cicd-as-platform.md]: Screenshot 2020-12-14 at 15 25 59

**Review by darren-reddick — APPROVED** [review]

**Comment by darren-reddick** [comment]
Have run it using a jekyll server locally and LGTM

**Pull request merged**